### PR TITLE
TRITON-2130 Add snapshot-limit to VMAPI

### PR DESCRIPTION
**Pull request by danmcd**
Very much a WIP.  I merely searched for ZFS vmadm(1M) tunables already in VMAPI and followed their lead.

### CODE REVIEW

File: docs/index.md [hunk 1]
```diff
@@ -98,6 +98,7 @@ update time.
...
 | zfs_io_priority          | Number                        | ZFS IO Priority                                                                                                                                                                                                           | Yes                 | Yes    | Yes    |
 | zlog_max_size            | Number                        | Sets the maximum size of the stdio.log file for a docker zone before rotation. NOTE: To be used by sdc-docker only.                                                                                                       | No                  | Yes    | Yes    |
+| zfs_snapshot_limit       | Number                        | Sets the number of ZFS snapshots a VM can take.                                                                                            | Yes                 | Yes    | Yes    |
```
Comment (trentm) [inline docs/index.md:101]: move this up a line to be in alphabetical order, please

File: lib/workflows/provision.js [hunk 2]
```diff
@@ -165,7 +165,7 @@ function preparePayload(job, cb) {
         'max_swap', 'mdata_exec_timeout', 'nics',
         'owner_uuid', 'quota', 'ram',
         'resolvers', 'vcpus', 'zfs_data_compression', 'zfs_io_priority',
-        'zlog_max_size', 'tags', 'tmpfs'
+	'zfs_snapshot_limit', 'zlog_max_size', 'tags', 'tmpfs'
```
Comment (trentm) [inline lib/workflows/provision.js:168]: Tab here.

File: docs/index.md (hunk 1 above)
Comment (danmcd) [reply docs/index.md:101]: Damn, off-by-one.

**Comment by danmcd** [comment]
>     * Are there valid ranges for this number? 0? Negative?

According to vmadm(1M):

`           type: integer (0+, set to '' or undefined to unset)`

>     * Should Triton clamp to a maximum value?

Good question.  I'd suggest something between 1000-9999.

>     * Should Triton start setting a default top value of 1000 or whatever?

For now, no.  This was requested by a customer (see Jira for more).

>     * Should that value be controllable via SAPI config var?

NO idea.

>     * Is there a proposal to expose this to cloudapi and node-triton? I don't think that necessarily needs to be true if the only immediate product requirement is to allow an operator to put a sane limit on this.

Currently no such proposal exists.  I have limited data from the SWSUP ticket that was filed.  Please visit there for more details (and I'll put your questions there, too).

File: lib/workflows/provision.js (hunk 2 above)
Comment (danmcd) [reply lib/workflows/provision.js:168]: Doh!  This isn't illumos-gate...

Comment (danmcd) [reply lib/workflows/provision.js]: Pushed fix.

**Comment by danmcd** [comment]
Manual tests have succeeded.  Plumbs cleanly all the way down to vmadm(1M) and the zone/VM itself.  See Jira for details.

**Comment by danmcd** [comment]
It appears the manual tests I ran cannot *easily* be put into the tests/ directory.

**Comment by danmcd** [comment]
I'll file a followup.

**Pull request merged**